From 894c39d798534fd06ad4e423361c0840017162fc Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Tue, 26 Jul 2022 02:22:54 +0000
Subject: [PATCH] xen/arm: mpam: Probe the MPAM hardware features
 supported

MPAM offers a series of hardware features that can be used in
different scenarios, including:
- Cache control features such as cache portion partitioning and
  cache capacity partitioning.
- Memory bandwidth control features such as memory bandwidth
  partitioning and max/min memory bandwidth partitioning.
- Performance monitoring features such as cache/membw monitor.
etc.

This commit expands the MPAM hardware probing support with the
feature probing. This is essential for resetting the MPAM controls
(in following patches) at the MPAM CPU online stage.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-5049
---
 xen/arch/arm/include/asm/mpam.h |  57 ++++++++++++++
 xen/arch/arm/mpam.c             | 129 ++++++++++++++++++++++++++++++++
 2 files changed, 186 insertions(+)

diff --git a/xen/arch/arm/include/asm/mpam.h b/xen/arch/arm/include/asm/mpam.h
index 8ce38d2358..1ae5cbecc8 100644
--- a/xen/arch/arm/include/asm/mpam.h
+++ b/xen/arch/arm/include/asm/mpam.h
@@ -266,6 +266,61 @@ enum mpam_class_types {
     MPAM_CLASS_UNKNOWN,     /* Everything else, e.g. SMMU */
 };
 
+/*
+ * When we compact the supported features, we don't care what they are.
+ * Storing them as a bitmap makes life easy.
+ */
+typedef uint16_t mpam_features_t;
+
+/* Bits for mpam_features_t */
+enum mpam_device_features {
+    mpam_feat_ccap_part = 0,
+    mpam_feat_cpor_part,
+    mpam_feat_mbw_part,
+    mpam_feat_mbw_min,
+    mpam_feat_mbw_max,
+    mpam_feat_mbw_prop,
+    mpam_feat_intpri_part,
+    mpam_feat_intpri_part_0_low,
+    mpam_feat_dspri_part,
+    mpam_feat_dspri_part_0_low,
+    mpam_feat_msmon,
+    mpam_feat_msmon_csu,
+    mpam_feat_msmon_csu_capture,
+    mpam_feat_msmon_mbwu,
+    mpam_feat_msmon_mbwu_capture,
+    mpam_feat_msmon_capt,
+    mpam_feat_partid_nrw,
+    MPAM_FEATURE_LAST,
+};
+#define MPAM_ALL_FEATURES    ((1<<MPAM_FEATURE_LAST) - 1)
+
+struct mpam_props
+{
+    mpam_features_t features;
+
+    uint16_t cpbm_wd;
+    uint16_t mbw_pbm_bits;
+    uint8_t bwa_wd;
+    uint16_t cmax_wd;
+    uint16_t intpri_wd;
+    uint16_t dspri_wd;
+    uint16_t num_csu_mon;
+    uint16_t num_mbwu_mon;
+};
+
+static inline bool mpam_has_feature(enum mpam_device_features feat,
+                                    struct mpam_props *props)
+{
+    return (1 << feat) & props->features;
+}
+
+static inline void mpam_set_feature(enum mpam_device_features feat,
+                                    struct mpam_props *props)
+{
+    props->features |= (1 << feat);
+}
+
 struct msc_feature_page
 {
     /* Physical addr defined in firmware. */
@@ -324,6 +379,8 @@ struct mpam_component
 struct mpam_msc_ris
 {
     uint8_t ris_idx;
+    uint64_t idr;
+    struct mpam_props props;
     cpumask_t affinity;
     /* member of mpam_component:ris */
     struct list_head comp_list;
diff --git a/xen/arch/arm/mpam.c b/xen/arch/arm/mpam.c
index 7102e2bc06..0025a2bdfc 100644
--- a/xen/arch/arm/mpam.c
+++ b/xen/arch/arm/mpam.c
@@ -685,6 +685,128 @@ static void __mpam_part_sel(uint8_t ris_idx, uint16_t partid,
     mpam_write_reg(msc, MPAMCFG_PART_SEL, partsel);
 }
 
+static void __init mpam_ris_hw_probe(struct mpam_msc_ris *ris)
+{
+    struct mpam_msc *msc = ris->msc;
+    struct mpam_props *props = &ris->props;
+    struct mpam_class *class = ris->comp->class;
+
+    BUG_ON(!spin_is_locked(&msc->lock));
+    BUG_ON(!spin_is_locked(&msc->hw_lock));
+
+    /* Cache Capacity Partitioning */
+    if ( FIELD_GET(MPAMF_IDR_HAS_CCAP_PART, ris->idr) )
+    {
+        uint32_t ccap_features = mpam_read_reg(msc, MPAMF_CCAP_IDR);
+        props->cmax_wd = FIELD_GET(MPAMF_CCAP_IDR_CMAX_WD, ccap_features);
+        if ( props->cmax_wd )
+            mpam_set_feature(mpam_feat_ccap_part, props);
+    }
+
+    /* Cache Portion partitioning */
+    if ( FIELD_GET(MPAMF_IDR_HAS_CPOR_PART, ris->idr) )
+    {
+        uint32_t cpor_features = mpam_read_reg(msc, MPAMF_CPOR_IDR);
+
+        props->cpbm_wd = FIELD_GET(MPAMF_CPOR_IDR_CPBM_WD, cpor_features);
+        if ( props->cpbm_wd )
+            mpam_set_feature(mpam_feat_cpor_part, props);
+    }
+
+    /* Memory bandwidth partitioning */
+    if ( FIELD_GET(MPAMF_IDR_HAS_MBW_PART, ris->idr) )
+    {
+        uint32_t mbw_features = mpam_read_reg(msc, MPAMF_MBW_IDR);
+
+        /* portion bitmap resolution */
+        props->mbw_pbm_bits = FIELD_GET(MPAMF_MBW_IDR_BWPBM_WD, mbw_features);
+        if ( props->mbw_pbm_bits &&
+             FIELD_GET(MPAMF_MBW_IDR_HAS_PBM, mbw_features) )
+            mpam_set_feature(mpam_feat_mbw_part, props);
+
+        props->bwa_wd = FIELD_GET(MPAMF_MBW_IDR_BWA_WD, mbw_features);
+        if ( props->bwa_wd && FIELD_GET(MPAMF_MBW_IDR_HAS_MAX, mbw_features) )
+            mpam_set_feature(mpam_feat_mbw_max, props);
+
+        if ( props->bwa_wd && FIELD_GET(MPAMF_MBW_IDR_HAS_MIN, mbw_features) )
+            mpam_set_feature(mpam_feat_mbw_min, props);
+
+        if ( props->bwa_wd && FIELD_GET(MPAMF_MBW_IDR_HAS_PROP, mbw_features) )
+            mpam_set_feature(mpam_feat_mbw_prop, props);
+    }
+
+    /* Priority partitioning */
+    if ( FIELD_GET(MPAMF_IDR_HAS_PRI_PART, ris->idr) )
+    {
+        uint32_t pri_features = mpam_read_reg(msc, MPAMF_PRI_IDR);
+
+        props->intpri_wd = FIELD_GET(MPAMF_PRI_IDR_INTPRI_WD, pri_features);
+        if ( props->intpri_wd &&
+             FIELD_GET(MPAMF_PRI_IDR_HAS_INTPRI, pri_features) )
+        {
+            mpam_set_feature(mpam_feat_intpri_part, props);
+            if ( FIELD_GET(MPAMF_PRI_IDR_INTPRI_0_IS_LOW, pri_features) )
+                mpam_set_feature(mpam_feat_intpri_part_0_low, props);
+        }
+
+        props->dspri_wd = FIELD_GET(MPAMF_PRI_IDR_DSPRI_WD, pri_features);
+        if ( props->dspri_wd &&
+             FIELD_GET(MPAMF_PRI_IDR_HAS_DSPRI, pri_features) )
+        {
+            mpam_set_feature(mpam_feat_dspri_part, props);
+            if ( FIELD_GET(MPAMF_PRI_IDR_DSPRI_0_IS_LOW, pri_features) )
+                mpam_set_feature(mpam_feat_dspri_part_0_low, props);
+        }
+    }
+
+    /* Performance Monitoring */
+    if ( FIELD_GET(MPAMF_IDR_HAS_MSMON, ris->idr) )
+    {
+        uint32_t msmon_features = mpam_read_reg(msc, MPAMF_MSMON_IDR);
+
+        if ( FIELD_GET(MPAMF_MSMON_IDR_MSMON_CSU, msmon_features) )
+        {
+            uint32_t csumonidr;
+
+            if ( !msc->nrdy_us )
+                printk(XENLOG_ERR
+                       "MPAM: Counters not usable because MSC%d not-ready timeout was not provided by firmware\n",
+                       msc->identifier);
+
+            csumonidr = mpam_read_reg(msc, MPAMF_CSUMON_IDR);
+            props->num_csu_mon = FIELD_GET(MPAMF_CSUMON_IDR_NUM_MON, csumonidr);
+            if ( props->num_csu_mon && msc->nrdy_us )
+                mpam_set_feature(mpam_feat_msmon_csu, props);
+        }
+        if ( FIELD_GET(MPAMF_MSMON_IDR_MSMON_MBWU, msmon_features) )
+        {
+            uint32_t mbwumonidr = mpam_read_reg(msc, MPAMF_MBWUMON_IDR);
+
+            props->num_mbwu_mon = FIELD_GET(MPAMF_MBWUMON_IDR_NUM_MON,
+                                            mbwumonidr);
+            if ( props->num_mbwu_mon )
+                mpam_set_feature(mpam_feat_msmon_mbwu, props);
+        }
+    }
+
+    /*
+     * RIS with PARTID narrowing don't have enough storage for one
+     * configuration per PARTID. If these are in a class we could use,
+     * reduce the supported partid_max to match the numer of intpartid.
+     * If the class is unknown, just ignore it.
+     */
+    if ( FIELD_GET(MPAMF_IDR_HAS_PARTID_NRW, ris->idr) &&
+         class->type != MPAM_CLASS_UNKNOWN )
+    {
+        uint32_t nrwidr = mpam_read_reg(msc, MPAMF_PARTID_NRW_IDR);
+        uint16_t partid_max = FIELD_GET(MPAMF_PARTID_NRW_IDR_INTPARTID_MAX,
+                                        nrwidr);
+
+        mpam_set_feature(mpam_feat_partid_nrw, props);
+        msc->partid_max = min(msc->partid_max, partid_max);
+    }
+}
+
 static int __init mpam_msc_hw_probe(struct mpam_msc *msc)
 {
     uint64_t idr;
@@ -728,6 +850,13 @@ static int __init mpam_msc_hw_probe(struct mpam_msc *msc)
         ris = mpam_get_or_create_ris(msc, ris_idx);
         if ( IS_ERR(ris) )
             return PTR_ERR(ris);
+
+        ris->idr = idr;
+
+        spin_lock_irqsave(&msc->hw_lock, flags);
+        __mpam_part_sel(ris_idx, 0, msc);
+        mpam_ris_hw_probe(ris);
+        spin_unlock_irqrestore(&msc->hw_lock, flags);
     }
 
     ACCESS_ONCE(mpam_partid_max) = min(ACCESS_ONCE(mpam_partid_max),
-- 
2.25.1

